python/veles/db/tracker.py

```diff
@@ -0,0 +1,349 @@
+# Copyright 2017 CodiLime
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import weakref
+
+try:
+    from functools import lru_cache
+except ImportError:
+    from backports.functools_lru_cache import lru_cache
+
+from veles.schema.nodeid import NodeID
+from veles.proto import operation, check
+from veles.proto.node import Node, PosFilter
+from veles.proto.exceptions import (
+    ObjectGoneError,
+    ObjectExistsError,
+    ParentCycleError,
+    PreconditionFailedError,
+)
+
+from .subscriber import (
+    BaseSubscriberNode,
+    BaseSubscriberData,
+    BaseSubscriberBinData,
+    BaseSubscriberList,
+)
+from .backend import DbBackend
+from .transaction import Transaction
+from .node import DbNode
+
+
+class DbTracker(object):
+    def __init__(self, db):
+        if not isinstance(db, DbBackend):
+            db = DbBackend(db)
+        self.db = db
+        # all_subs is a dict mapping all non-query subscriptions in existence
+        # to their DbNodes - this exists
+        # so that all involved objects are strongly referenced (otherwise,
+        # being only reacheble through the weak dict, they could be GCd along
+        # with all their subscriptions).
+        self.all_subs = {}
+        # WTF H4X: this strange construction does two things:
+        #
+        # - the self.nodes dictionary ensures there can be at most one DbNode
+        #   for every node, so there is only one place to invalidate.
+        # - the lru_cache on get_cached_node ensures at least 128 last nodes
+        #   are actually kept alive.
+        self.nodes = weakref.WeakValueDictionary()
+        self.get_cached_node = lru_cache(maxsize=128)(self._get_cached_node)
```

Comment:
Hmm, maybe make this 128 a parameter? Since it will probably one of parameters to fine-tuning the server.